File: C++/687. Longest Univalue Path.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int dfs(TreeNode* node, int val)
    {
        if(node == nullptr || node->val != val)
            return 0;
        else
            return max(dfs(node->left, val),dfs(node->right, val)) + 1;
    }
    int longestUnivaluePath(TreeNode* root) {
        if(root == nullptr)
            return 0;
        int ans = dfs(root->left, root->val)+dfs(root->right, root->val);
        int res = max(longestUnivaluePath(root->right),longestUnivaluePath(root->left));
        return max(ans, res);
    }
};

File: Python/687. Longest Univalue Path.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def longestUnivaluePath(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        def dfs1(val, node):
            if node == None or node.val != val:
                return 0
            else:
                return max(dfs1(val, node.left), dfs1(val, node.right)) + 1
        def dfs2(node):
            if node == None:
                return 0
            ans = dfs1(node.val, node.left) + dfs1(node.val, node.right)
            res = max(dfs2(node.left), dfs2(node.right))
            return max(ans, res)
        return dfs2(root)